### PR TITLE
Ingela/ssl/default cert sign/gh 8588/otp 19152

### PR DESCRIPTION
**Pull request by IngelaAndin**
Smoother upgrade path, allow certs to still use sha1 but not protocol signatures.

### CODE REVIEW

File: lib/ssl/test/ssl_api_SUITE.erl [hunk 1]
```diff
@@ -597,6 +600,49 @@ select_sha1_cert(Config) when is_list(Config) ->
                                                    {key, {namedCurve, secp256r1}}]}}),
     test_sha1_cert_conf(Version, TestConfRSA, TestConfECDSA, Config).
 
+%%--------------------------------------------------------------------
+root_any_sign() ->
+    [{doc,"Use cert signed with unsported signature for the root will suceed, as it is not verified"}].
```
Comment (u3s) [inline lib/ssl/test/ssl_api_SUITE.erl:605]: 2 typos:
`unsported -> unsupported`
`suceed -> succeed`

File: lib/ssl/test/ssl_api_SUITE.erl [hunk 2]
```diff
@@ -597,6 +600,49 @@ select_sha1_cert(Config) when is_list(Config) ->
...
+
+root_any_sign(Config) when is_list(Config) ->
+    Version = ssl_test_lib:protocol_version(Config),
+    #{client_config := CSucess, server_config := SSucess} =
```
Comment (u3s) [inline lib/ssl/test/ssl_api_SUITE.erl:609]: maybe write a code comment in line 609 and 624 to explain what is relevant in those configs from test perspective;
please explain why 2nd one is expected to fail

File: lib/ssl/test/ssl_api_SUITE.erl (hunk 1 above)
Comment (IngelaAndin) [reply lib/ssl/test/ssl_api_SUITE.erl:605]: fixed

File: lib/ssl/test/ssl_api_SUITE.erl (hunk 2 above)
Comment (IngelaAndin) [reply lib/ssl/test/ssl_api_SUITE.erl:609]: fixed

File: lib/ssl/src/ssl_handshake.erl [hunk 3]
```diff
@@ -1481,10 +1482,11 @@ signature_algs_ext(SignatureSchemes0) ->
 
 signature_algs_cert(undefined) ->
     undefined;
+signature_algs_cert([default| SignatureSchemes]) ->
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl_handshake.erl:1485]: fixed

File: lib/ssl/src/dtls_handshake.erl [hunk 4]
```diff
@@ -175,16 +175,20 @@ handle_client_hello(Version,
...
 		    #{versions := Versions,
-                      signature_algs := SupportedHashSigns,
                       eccs := SupportedECCs,
                       honor_ecc_order := ECCOrder} = SslOpts,
 		    {SessIdTracker, Session0, ConnectionStates0, CertKeyPairs, _},
                     Renegotiation) ->
     case dtls_record:is_acceptable_version(Version, Versions) of
 	true ->
+            TLSVersion = dtls_v1:corresponding_tls_version(Version),
+             SupportedHashSigns =
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/dtls_handshake.erl:185]: fixed

File: lib/ssl/src/ssl.erl [hunk 5]
```diff
@@ -4121,29 +4125,82 @@ server_name_indication_default(_) ->
...
+
+opt_signature_algs_not_valid(UserOpts, #{versions := Versions} = Opts0)->
+    Opts =
+        case get_opt_list(signature_algs, undefined, UserOpts, Opts0) of
+            {default, undefined} ->
+                Opts0#{signature_algs => undefined};
+            {old, _} ->
+                Opts0;
+            _ ->
+                assert_version_dep(signature_algs, Versions, ['tlsv1.2', 'tlsv1.3']),
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl.erl:4187]: Yes I looked into this and I think it came to be from flaky tests code. I have changed it.

File: lib/ssl/src/ssl.erl [hunk 6]
```diff
@@ -407,8 +407,12 @@ extension is not need.
 
 The client will send a `signature_algorithms_cert` extension (in the client
 hello message), if TLS version 1.2 (back-ported to TLS 1.2 in 24.1) or later is
-used, and the signature_algs_cert option is explicitly specified. By default,
-only the [signature_algs](`t:signature_algs/0`) extension is sent.
+used, and the signature_algs_cert option is explicitly specified.
+By default, only the [signature_algs](`t:signature_algs/0`) extension is sent with the
+exeption of when signature_algs option is not explicitly specified, in which case it
+will use the default value for signature_algs plus rsa_pkcs1_sha1 to allow
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl.erl:413]: Rephrased

File: lib/ssl/src/ssl.erl [hunk 7]
```diff
@@ -407,8 +407,12 @@ extension is not need.
 
 The client will send a `signature_algorithms_cert` extension (in the client
 hello message), if TLS version 1.2 (back-ported to TLS 1.2 in 24.1) or later is
-used, and the signature_algs_cert option is explicitly specified. By default,
-only the [signature_algs](`t:signature_algs/0`) extension is sent.
+used, and the signature_algs_cert option is explicitly specified.
+By default, only the [signature_algs](`t:signature_algs/0`) extension is sent with the
+exeption of when signature_algs option is not explicitly specified, in which case it
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/src/ssl.erl:412]: fixed

File: lib/ssl/test/ssl_cert_SUITE.erl [hunk 8]
```diff
@@ -1224,7 +1224,7 @@ unsupported_sign_algo_cert_client_auth(Config) ->
         'tlsv1.3' ->
             ssl_test_lib:basic_alert(ClientOpts, ServerOpts, Config, certificate_required);
         _  ->
-            ssl_test_lib:basic_alert(ClientOpts, ServerOpts, Config, unsupported_certificate)
+            ssl_test_lib:basic_alert(ClientOpts, ServerOpts, Config, insufficient_security)
```
[the remark this answers is not in the document]
Comment (IngelaAndin) [reply lib/ssl/test/ssl_cert_SUITE.erl:1227]: I thought I had an explanation, but it seems it was the flaky tests that was somehow responsible for this. Changed back now.

**Review by u3s — APPROVED** [review]

**Pull request merged**